python/pyarrow/pandas_compat.py

```diff
@@ -310,16 +310,30 @@ def dataframe_to_arrays(df, schema, preserve_index):
             field = schema.field_by_name(name)
             type = getattr(field, "type", None)
 
-        array = pa.array(col, from_pandas=True, type=type)
-        arrays.append(array)
+        columns_to_convert.append(col)
+        convert_types.append(type)
         names.append(name)
-        types.append(array.type)
 
     for i, column in enumerate(index_columns):
-        array = pa.array(column)
-        arrays.append(array)
+        columns_to_convert.append(column)
+        convert_types.append(None)
         names.append(index_level_name(column, i))
-        types.append(array.type)
+
+    def convert_column(col, ty):
+        return pa.array(col, from_pandas=True, type=ty)
+
+    if nthreads == 1:
+        arrays = [convert_column(c, t)
+                  for c, t in zip(columns_to_convert,
+                                  convert_types)]
+    else:
+        from concurrent import futures
+        with futures.ThreadPoolExecutor(nthreads) as executor:
```

Comment:
Is it much slower to just use this code path?

Reply:
I doubt it, it was more the principle of starting a thread pool for no reason. I'd be find with this being the only code path

Reply:
The thread pool seems to have some non-trivial fixed overhead, at least 20 microseconds per task on my machine. I did some quick testing and when the number of columns is "large" relative to the number of rows, then using a thread pool is slower than single thread. This suggests we should use some heuristic to decide whether to use the thread pool to avoid bad performance in wide tables without a ton of rows.

For example

```
NROWS = 10000
NCOLS = 500

arr = np.random.randn(NCOLS, NROWS).T
arr[::5] = np.nan

df = pd.DataFrame(arr)

%timeit sdf = pa.serialize_pandas(df, nthreads=1)
10 loops, best of 3: 62.8 ms per loop

%timeit sdf = pa.serialize_pandas(df, nthreads=2)
10 loops, best of 3: 83.4 ms per loop
```

Reply:
Added a rough heuristic to turn off the thread pool if number of rows is less than 100 times the number of columns